### PR TITLE
Generate symbol table when creating relocations in binary-writer.cc

### PR DESCRIPTION
**Pull request by sbc100**

### CODE REVIEW

File: src/binary-reader-objdump.cc [hunk 1]
```diff
@@ -134,24 +135,29 @@ const char* BinaryReaderObjdumpBase::GetGlobalName(Index index) const {
...
+  ObjdumpSymbol& sym = objdump_state_->symtab[symbol_index];
+  switch (sym.kind) {
+    case SymbolType::Function:
+      return GetFunctionName(sym.index);
+    case SymbolType::Data:
+      return sym.name;
+    case SymbolType::Global:
+      return GetGlobalName(sym.index);
+  }
+}
```
Comment (binji) [inline src/binary-reader-objdump.cc]: gcc gives an error here: `error: control reaches end of non-void function`

File: src/binary-reader-objdump.cc [hunk 2]
```diff
@@ -1241,7 +1292,9 @@ Result BinaryReaderObjdump::OnGlobalSymbol(Index index,
                                            Index global_index) {
   std::string sym_name = name.to_string();
   if (sym_name.empty()) {
-    sym_name = GetGlobalName(global_index);
+    if (const char* N = GetGlobalName(global_index)) {
```
Comment (binji) [inline src/binary-reader-objdump.cc]: nit: gotta be something better than `N` here :-)

File: src/binary-reader-objdump.h [hunk 3]
```diff
@@ -50,13 +50,20 @@ struct ObjdumpOptions {
   const char* section_name;
 };
 
+struct ObjdumpSymbol {
+  wabt::SymbolType kind;
+  string_view name;
```
Comment (binji) [inline src/binary-reader-objdump.h]: Maybe `std::string` here? Who owns the string_view memory?

File: src/binary-writer.cc [hunk 4]
```diff
@@ -289,6 +298,30 @@ Index BinaryWriter::GetExceptVarDepth(const Var* var) {
...
+      break;
+    case RelocType::GlobalIndexLEB:
+      type = SymbolType::Global;
+      name = module_->globals[index]->name;
+      break;
+    default:
+      WABT_UNREACHABLE;
+  }
+  auto iter = symtab_.find(name);
+  if (iter != symtab_.end())
```
Comment (binji) [inline src/binary-writer.cc]: nit: braces

File: src/binary-writer.cc [hunk 5]
```diff
@@ -289,6 +298,30 @@ Index BinaryWriter::GetExceptVarDepth(const Var* var) {
...
+    return iter->second;
+
+  Index sym_index = Index(symbols_.size());
+  symtab_[name] = sym_index;
+  symbols_.push_back(Symbol{sym_index, type, index});
```
Comment (binji) [inline src/binary-writer.cc]: `emplace_back` instead?

File: src/ir.h [hunk 6]
```diff
@@ -133,7 +133,6 @@ struct FuncSignature {
 };
 
 struct FuncType {
-  FuncType() = default;
```
Comment (binji) [inline src/ir.h]: hm, not sure how I feel about this change. I don't think these should always have names.

Comment (sbc100) [reply src/ir.h]: But all the callsites to provide a name.   No point in having non-used constructor is there?

File: src/binary-writer.cc (hunk 5 above)
Comment (sbc100) [reply src/binary-writer.cc]: I don't think that works with POD structs

File: src/binary-reader-objdump.h (hunk 3 above)
Comment (sbc100) [reply src/binary-reader-objdump.h]: good point.  I was assuming it was the binary file, but I guess the strings that are provided in the reader callbacks don't outlive the reader itself.

File: src/ir.h (hunk 6 above)
Comment (binji) [reply src/ir.h]: Guess so.

**Pull request merged**